loudml/prometheus.py

```diff
@@ -0,0 +1,397 @@
+# bottomk (smallest k elements by sample value)
+# topk (largest k elements by sample value)
+# quantile (calculate φ-quantile (0 ≤ φ ≤ 1) over dimensions)
+
+AGGREGATORS = {
+    'avg': 'avg({}{})',
+    'mean': 'avg({}{})',
+    'average': 'avg({}{})',
+    'stddev': 'stddev({}{})',
+    'std_dev': 'stddev({}{})',
+    'count': 'count({}{})',
+    'min': 'min({}{})',
+    'max': 'max({}{})',
+    'sum': 'sum({}{})',
+    'bottomk': 'bottomk(1,{}{})', # TODO: need to handle bottomk(X) param
+    'topk': 'topk(1,{}{})', # TODO: need to handle topk(X) param
+    'deriv': None,
+    'derivative': None,
+    'integral': None,
+    'med': None,
+    'median': None,
+    'mode': None,
+    '5percentile': 'quantile(0.05,{}{})',
+    '10percentile': 'quantile(0.10,{}{})',
+    '90percentile': 'quantile(0.90,{}{})',
+    '95percentile': 'quantile(0.95,{}{})'
+}
+
+
+def _build_tags_predicates(match_all=None):
+    """
+    Build tags predicates
+    """
+    must = []
+
+    if match_all:
+        for condition in match_all:
+            must.append('{}="{}"'.format(condition['tag'], condition['value']))
+
+    return "{" + ",".join(must) + "}"
+
+
+class PrometheusResult(object):
+    """
+    Helper class to parse query result
+    """
+
+    def __init__(self, response):
+        self._response = response
+
+    def __repr__(self):
+        return "Prometheus results: {}...".format(str(self._response)[:200])
+
+    def get_points(self):
+        if ((not self._response)
+            or (not 'data' in self._response)
+            or (not 'result' in self._response['data'])
+            or (not len(self._response['data']['result']))
+        ):
+            return []
+        # values are pairs of [timestamp, value]
+        return self._response['data']['result'][0]['values']
+
+
+class PrometheusClient(object):
+    """
+    Client for Prometheus
+    """
+
+    def __init__(
+        self,
+        host="localhost",
+        port=9090,
+        ssl=False,
+        verify_ssl=False,
+        ssl_cert_path='',
+        user='',
+        password=''
+    ):
+        """
+        Set proper schema based on SSL param, open session
+        and supply basic authentication creds if given
+        https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification
+        """
+        schema = "http"
+        if ssl:
+            schema = "https"
+
+        self.url = "%s://%s:%d" % (schema, host, port)
+        self.session = requests.Session()
+        if user and password:
+            self.session.auth = (user, password)
+        if ssl_cert_path:
+            self.session.verify = ssl_cert_path
+
+
+    def build_url_params(self, q):
+        """
+        Forms a query URL params from bits.
+        TODO: add better aggregator functions handling
+        """
+        params = {
+            'start': int(make_ts(q['start'])),
+            'end': int(make_ts(q['end'])),
+            'step': q["step"]
```

Comment:
The API docs mention a timeout. Should we try to set it this field in params?

```
query=<string>: Prometheus expression query string.
time=<rfc3339 | unix_timestamp>: Evaluation timestamp. Optional.
timeout=<duration>: Evaluation timeout. Optional. Defaults to and is capped by the value of the -query.timeout flag.
```

timeout: DEFAULT_REQUEST_TIMEOUT